Update CCEffectHue with color matrix implementation. Fix compiler warnings.

thayerandrews opened the pull request:
- As Scott pointed out in his review of my previous pull request, the hue effect could be greatly optimized by using a 4x4 color matrix to do the RGB to YIQ conversion, rotation, and conversion back to RGB. With this optimization, performance of the hue effect test on my iPhone 4s increased from 26fps to 60fps. The original implementation is still in place and disabled with a #if but I will remove it in the future after more testing of the color matrix implementation.
- Remove a bunch of strong reference cycles in the effect pass blocks.
- Fix other miscellaneous compiler warnings from the OSX build.

Commit:
CCEffects - Fix compiler warnings

- CCShaderUniformPreviousPassTexture should be "NSString * const" not "const NSString *". The former
gets rid of a compiler warning when passing the constant as a parameter and it also matches string constant
declarations in Apple's APIs. Should we update all the other shader constants (and maybe all constants
everywhere)? If so that's another change for another day.

- Stop using the deprecated version of [CCRenderState initWithBlendMode...]

Commit:
CCEffects - Fix compiler warnings on OSX

- One of this is an unused variable warning so that's just commented out.
- One was a legitimate bug caused by a copy-paste error.

Commit:
CCEffects - Fix compiler warning and add an assert for an unsupported situation

We don't currently support effects with multiple inputs* so assert that we're not getting
them. Because of this limitation we only need one temporary variable when stitching effects
together. Fix a compiler warning where we had a loop over all inputs that we don't need.

*Note: In this context, the word "input" means some color value that is potentially the output
of a preceding effect in an effect stack.

Commit:
CCEffects - Fix compiler warnings / strong reference cycles

Refer to the uniform translation table via weakSelf not self. This fixes a bunch of strong reference
cycles and also fixes a couple unused variable warnings for weakSelf.

Commit:
CCEffectHue - Add a color matrix implementation as suggested by Scott

Commit:
CCEffectHue - Recompute the color matrix as needed instead of on every draw

This was motivated by the following performance findings on an iPhone 4s.

Original implementation: 26fps / 20% CPU usage
Color matrix implementation: `60fps / 44% CPU usage

Interestingly even removing the matrix computation on every draw, the CPU usage is still
~44%. Since the change doesn't add much complexity I'm keeping it but I'm surprised it
had no impact on CPU usage.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
CCEffectRefraction - Disable the refraction effect according to the normal map's alpha channel

If alpha == 0, then you just get the base color and texture but no contribution from the environment.
This is implemented as a multiply not a conditional so alpha values between 0 and 1 will modulate
the environment color.

Commit:
CCEffectsTest - Update torus images

slembcke commented on cocos2d/CCEffectHue.m:
```diff
@@ -157,4 +192,32 @@ float conditionHue(float hue)
+    // in the shader instead of mtx * color.  The order that the matrices are composed below is
+    // also reversed to account for the left multiply of the color value.
+    
+    GLKMatrix4 rgbToYiq = GLKMatrix4Make(0.299,     0.587,     0.114,    0.0,
```
This is a little pedantic, but why not use GLKMatrix4MakeAndTranspose() instead? Then you don't have to reverse the multiplication order nor mix transposed and non-transposed matrices. Not technically wrong, but it does make it non-obvious to follow along without the comment explaining that you purposefully transposed the color transform matrices.

Also, since the untransposed rotation matrix is implicitly inverted as a special case for y/z rotations, I wonder if the hue angle should be negated or not? Especially with the comment in the old code that questioned if the hue was backwards or not.

thayerandrews replied:
Haha. My initial implementation of the color matrix used GLKMatrix4MakeAndTranspose() instead and everything else was reversed. I'll switch it back.

slembcke replied:
:p Ok. I was kind of hesitant to bring it up since it's sort of a dumb nitpick.

thayerandrews replied:
Not at all. I went back and forth trying to figure out which way would be more obvious and in the end I think I basically chose at random. 

You're probably right that creating the matrices with GLKMatrix4MakeAndTranspose() is clearer since it keeps all the matrix operations in a more expected order (at least I feel like v' = Mv is more "standard" than v' = vM) and it puts the word "transpose" right there at matrix creation time.

thayerandrews replied:
Oh also, the sign of the hue rotation is interesting. If you look at this:

http://upload.wikimedia.org/wikipedia/commons/8/82/YIQ_IQ_plane.svg

And this:

http://upload.wikimedia.org/wikipedia/commons/5/52/HSL-HSV_hue_and_chroma.svg

You can see that rotating in the positive direction in YIQ is the opposite of positive hue rotation in HSV. To double check this I wrote a quick test that does the following conversions:
- RGB directly to HSV using the method described [here](http://en.wikipedia.org/wiki/HSL_and_HSV#General_approach).
- RGB to YIQ using the matrix from the code.
- Compute H from YIQ as the GPUImage source does.

```
rgb:( 1.000  0.000  0.000)  yiq:( 0.299  0.596  0.211)  hsv:(   0.000  1.000  1.000)  H from YIQ: 19.543 
rgb:( 1.000  1.000  0.000)  yiq:( 0.886  0.321 -0.311)  hsv:(  60.000  1.000  1.000)  H from YIQ:-44.082
rgb:( 0.000  1.000  0.000)  yiq:( 0.587 -0.274 -0.523)  hsv:( 120.000  1.000  1.000)  H from YIQ:-117.707
rgb:( 0.000  1.000  1.000)  yiq:( 0.701 -0.596 -0.211)  hsv:( 180.000  1.000  1.000)  H from YIQ:-160.475
rgb:( 0.000  0.000  1.000)  yiq:( 0.114 -0.321  0.311)  hsv:( 240.000  1.000  1.000)  H from YIQ: 135.898
rgb:( 1.000  0.000  1.000)  yiq:( 0.413  0.274  0.523)  hsv:( -60.000  1.000  1.000)  H from YIQ: 62.302
```

From this you can see that the H values from YIQ match the image of YIQ space but they don't correspond to H angles for HSV and, most importantly for this discussion, they are backwards relative to one another.

Commit:
CCEffectHue - Address Scott's review comments regarding matrix ordering

Create transposed versions of the matrices and reverse the order of all operations using them.
The effect is the same but we decided that this would be easier for others to understand.  Change
the comments around a bit to explain what's going on.

thayerandrews commented:
Updated CCEffectHue to address Scott's comments. This also includes a change to CCEffectRefraction to address Viktor's request for masking refraction based on the normal map's alpha channel.

slembcke commented:
Looks good to me. No more matrix nitpicks. ;)
